### ISSUE TITLE
Getting a Script Error and having Zero Contracts when enabling Better Contracts on a Server

### ISSUE THREAD

**Issue by cwattyeso** [Mmtrx/FS22_BetterContracts#82]
Have a Dedicated Multiplayer Server and without Better Contracts enabled, contracts appear in the menu fine. As soon as I enable Better Contracts on the Server and go in game the contract screen is empty and as Server Admin I can't clear or refresh contracts and none of the the buttons for the functions appear on the screen:

![fsScreen_2022_12_20_16_57_48](https://user-images.githubusercontent.com/13921716/208723133-d73a1037-0b01-4b51-a5dc-e89ecdb43261.png)

When I check the Server Log File I see this error:

`2022-12-20 16:53 Error: Running LUA method 'update'.
H:/Multiplayer Server/FS22_BetterContracts/betterContracts.lua:711: attempt to perform arithmetic on local 'price' (a nil value)
2022-12-20 16:53  Missing filltype WHEAT Dairy
2022-12-20 16:53 LUA call stack:
  dataS/scripts/objects/SellingStation.lua (430) : printCallstack
  =H:/Multiplayer Server/FS22_BetterContracts/betterContracts.lua (710) : getFilltypePrice
  =H:/Multiplayer Server/FS22_BetterContracts/betterContracts.lua (646) : addMission
  =H:/Multiplayer Server/FS22_BetterContracts/scripts/gui.lua (309) : refresh
  dataS/scripts/gui/InGameMenuContractsFrame.lua (154) : updateList
  =H:/Multiplayer Server/FS22_BetterContracts/scripts/gui.lua (167) : superFunc
  dataS/scripts/gui/base/TabbedMenu.lua (614) : onFrameOpen
  dataS/scripts/gui/InGameMenu.lua (852) : onPageChange
  dataS/scripts/gui/elements/PagingElement.lua (90) : raiseCallback
  dataS/scripts/gui/base/TabbedMenu.lua (551) : setPage
  dataS/scripts/gui/elements/MultiTextOptionElement.lua (424) : raiseCallback
  dataS/scripts/gui/elements/MultiTextOptionElement.lua (409) : raiseClickCallback
  dataS/scripts/gui/elements/MultiTextOptionElement.lua (373) : onRightButtonClicked
  dataS/scripts/gui/elements/ScreenElement.lua (142) : inputRight
  dataS/scripts/gui/base/TabbedMenu.lua (579) : onPageNext
  dataS/scripts/gui/elements/ScreenElement.lua (206) : onPageNext
  dataS/scripts/gui/elements/GuiElement.lua (652) : inputEvent
  dataS/scripts/gui/base/Gui.lua (782) : inputEvent
  dataS/scripts/gui/base/Gui.lua (809) : notifyControls
  dataS/scripts/input/InputEvent.lua (100) : callback
  dataS/scripts/input/InputBinding.lua (3328) : notifyInput
  dataS/scripts/input/InputBinding.lua (3041) : updateInput
  dataS/scripts/main.lua (1202) : update
2022-12-20 16:53 Error: Running LUA method 'update'.
H:/Multiplayer Server/FS22_BetterContracts/betterContracts.lua:711: attempt to perform arithmetic on local 'price' (a nil value)
2022-12-20 16:53  Missing filltype WHEAT Dairy
2022-12-20 16:53 LUA call stack:
  dataS/scripts/objects/SellingStation.lua (430) : printCallstack
  =H:/Multiplayer Server/FS22_BetterContracts/betterContracts.lua (710) : getFilltypePrice
  =H:/Multiplayer Server/FS22_BetterContracts/betterContracts.lua (646) : addMission
  =H:/Multiplayer Server/FS22_BetterContracts/scripts/gui.lua (309) : refresh
  dataS/scripts/MessageCenter.lua (158) : callback
  dataS/scripts/fieldJobs/AbstractMission.lua (49) : publish
  dataS/scripts/fieldJobs/AbstractFieldMission.lua (60) : delete
  dataS/scripts/fieldJobs/HarvestMission.lua (45) : delete
  dataS/scripts/network/Client.lua (322) : delete
  dataS/scripts/network/ConnectionManager.lua (28) : func
2022-12-20 16:53 Error: Running LUA method 'packetReceived'.
H:/Multiplayer Server/FS22_BetterContracts/betterContracts.lua:711: attempt to perform arithmetic on local 'price' (a nil value)
2022-12-20 16:53  Missing filltype WHEAT Dairy
2022-12-20 16:53 LUA call stack:
  dataS/scripts/objects/SellingStation.lua (430) : printCallstack
  =H:/Multiplayer Server/FS22_BetterContracts/betterContracts.lua (710) : getFilltypePrice
  =H:/Multiplayer Server/FS22_BetterContracts/betterContracts.lua (646) : addMission
  =H:/Multiplayer Server/FS22_BetterContracts/scripts/gui.lua (309) : refresh
  dataS/scripts/gui/InGameMenuContractsFrame.lua (154) : updateList
  =H:/Multiplayer Server/FS22_BetterContracts/scripts/gui.lua (167) : superFunc
  dataS/scripts/gui/base/TabbedMenu.lua (614) : onFrameOpen
  dataS/scripts/gui/InGameMenu.lua (852) : onPageChange
  dataS/scripts/gui/elements/PagingElement.lua (90) : raiseCallback
  dataS/scripts/gui/base/TabbedMenu.lua (551) : setPage
  dataS/scripts/gui/elements/MultiTextOptionElement.lua (424) : raiseCallback
  dataS/scripts/gui/elements/MultiTextOptionElement.lua (142) : raiseClickCallback
  dataS/scripts/gui/base/TabbedMenu.lua (168) : setState
  dataS/scripts/gui/elements/GuiElement.lua (772) : raiseCallback
  dataS/scripts/gui/base/Gui.lua (907) : onOpen
  =H:/Multiplayer Server/FS22_additionalGameSettings/scripts/Settings.lua (555) : superFunc
  dataS/scripts/FSBaseMission.lua (4469) : changeScreen
  dataS/scripts/input/InputEvent.lua (100) : callback
  dataS/scripts/input/InputBinding.lua (3328) : notifyInput
  dataS/scripts/input/InputBinding.lua (3041) : updateInput
  dataS/scripts/main.lua (1202) : update
2022-12-20 16:53 Error: Running LUA method 'update'.
H:/Multiplayer Server/FS22_BetterContracts/betterContracts.lua:711: attempt to perform arithmetic on local 'price' (a nil value)`

If I remove Better Contracts from the server and restart, the contract screen correctly appears again and contracts can be accepted.

**Comment by Mmtrx** [comment]
Hm, is this also on Maypole? 
Log errors hint at an unknown price for „wheat dairy“ (?) 
Will look into this asap.

**Comment by cwattyeso** [comment]
Yeah it is on Maypole on the new server I started today to try to fix the contract issues I was having previously.

The issue on closer examination of the contracts screen without Better Contracts installed seems to be related to Harvesting Contracts having Wrong Equipment being offered for Borrowing and also Wrong Crop to Wrong Sellpoint. For Example one Harvesting Contract is offering the John Deere Cotton Harvester to do a Harvesting Contract for Wheat which it then wants delivered to the Dairy which doesn't accept Wheat.

I'm going to try deleting the Missions.xml from the save folder and letting the game generate a new one and see if that fixes the problem in the first instance. If so then I can attempt to add Better Contracts back to the server. Better Contracts was working on fine on the other server but there wasn't a lot of contracts due to an issue I've not tracked down to being caused by another mod which absolutely breaks maps if you ever use it.

**Comment by cwattyeso** [comment]
![fsScreen_2022_12_20_18_48_37](https://user-images.githubusercontent.com/13921716/208744028-bdfd40e6-bc9d-4869-a0eb-1942172fec2c.png)
![fsScreen_2022_12_20_18_49_08](https://user-images.githubusercontent.com/13921716/208744039-aa81e06c-48d9-410e-b3f9-34c590562ea6.png)
![fsScreen_2022_12_20_18_49_27](https://user-images.githubusercontent.com/13921716/208744051-7b8de007-aee2-4e66-bd29-816d206f7d34.png)

The base game without Better Contracts is picking the wrong equipment for crop type and then choosing destinations for delivery which don't accept that crop, this I think it what is breaking and causing the script error when I try to add Better Contracts to the save.

I've tried deleting the missions.xml and generating a new one but the same keeps happening. The game is selecting all the wrong equipment types for jobs.

**Comment by Mmtrx** [comment]
Hi,
ja, that's a problem with the games list of possible loan vehicles for each mission type. This list is generated at game start by reading in 1 ore more `missionVehicle.xml` files. Besides the vanilla one, mods, DLCs, maps can also have additional ones. I checked Maypole, that map doesn't have one. But e.g. the Platinum DLC (and BetterContracts 😉) has one. So if a given savegame is restarted with one of these mods (or DLCs) either added or left out, the ingame mission vehicles list gets out of sync.

Solution is, as you say, deleting `missions.xml` from the respective savegame directory on the dedi server, and restart the savegame. What do you mean by "generating a new one"?

**Comment by cwattyeso** [comment]
I tried generating a new missions.xml by deleting the one from the server and then restarting the server but that didn't work.

In the end to fix the problem what I had to do was log in as admin and manually accept then cancel all the harvesting contracts to get rid of them from the missions screen. I also did the same with any contracts that didn't show any equipment for loan, however these ones disappeared themselves after I accepted them so didn't need to cancel those. Guess they were truly broken contracts.

Once I had the mission list cleaned up of anything that look wrong, I performed a manual save on the server, logged out, stopped the server, uploaded the latest version of Better Contracts and enabled it and then restarted the server. This time when I logged in I could see the mission screen fine and had the key inputs on the bottom of the screen for the Better Contract opens (Refresh, Clear, Details etc). I went into the game menu, tabbed to the Better Contract options and set those as I want on the server. I performed a manual save at this point. I then went to the nearest sleep trigger and advanced the game to the next day. I opened the missions screen again and used the refresh contracts option to generate new Harvest Contracts that now have the correct equipment and also the correct delivery locations.
